### DISCUSSION TITLE
startReceive fails with code -2

### DISCUSSION THREAD

**Discussion by SamLue** [jgromes/RadioLib#1165] [Q&A] [answered]
Hi,

I get the error -2, when I try to listen with radio.startReceive()

First some Information about my setup:
I'm using a sx1262 (ra-01sh). Here some specs: https://docs.ai-thinker.com/_media/lora/docs/ra-01sh_specification.pdf

The Problem occurs, if I connect this Module with a ESP32 (WROOM).
Interesting: If I use this Module with the STM32F103CB, there is no Problem with listening.

The first Problem was, that I have already got the error -707 at .begin with the default tcxoVoltage 1.6. Only If I disable this tcxoVoltage to Value 0 the .begin ist successful.
With the STM32 and the same Module everything works with default tcxoVoltage 1.6

But OK, with this setting and the esp32, it is possible send Messages for example with:  transmissionState = radio.startTransmit("Hello World!");

But startReceive() fails with Code -2
If I check the Error-Codes: Radio chip was not found during initialization. This can be caused by specifying wrong chip type in the constructor (i.e. calling [SX1272](https://jgromes.github.io/RadioLib/class_s_x1272.html) constructor for [SX1278](https://jgromes.github.io/RadioLib/class_s_x1278.html) chip) or by a fault in your wiring (incorrect slave select pin).

chip was not found? I would assume this error at .begin but not at .startReceive!?
But how can I check "incorrect slave select pin"?

Here my PIN definition and wiring:
```
// MOSI pin:   23
// MISO  pin:  19
// SCK pin:  18
#define SX1262_NSS  5
#define SX1262_DIO1 12
#define SX1262_NRST 14
#define SX1262_BUSY 4
#define SX1262_TCXOVOLTAGE 0
#endif

SX1262 radio = new Module(SX1262_NSS, SX1262_DIO1, SX1262_NRST, SX1262_BUSY);

```
An here the error with debug output enabled:

```
[SX1262] Starting to listen ... RLB_SPI: CMDW   8
RLB_SPI: SI             2       62      0       2       0       0       0       0
RLB_SPI: SO     FF      FF      FF      FF      FF      FF      FF      FF      FF
failed, code -2
```

Any Idea, how to solve this error?

**Comment by jgromes** [comment]
> With the STM32 and the same Module everything works with default tcxoVoltage 1.6

That is very strange - the SX1268 modules either has a TCXO, and then it needs a non-zero value for `tcxoVoltage` or it does have an XTAL and then it needs a zero value for `tcxoVoltage`. If it is set wrong the `begin` method should fail, so this already points to some problem, most likely in the hardware.

> I would assume this error at .begin 

That is indeed where it comes from most of the time. However, if at some point after `begin`, the SPI starts returning 0xFF or 0x00 instead of any meaningful data, this cannot be distinguished from an unconnected radio module. In these cases, any SX126x method will return that error.

You haven't posted your code, so it's impossible to tell what is actually going on. Though seeig the erratic TCXO behavior my first guess would be some hardware issue, for example insufficient power supply.

**Reply by SamLue** [reply]
The ra-01sh seems to be a sx1262 (here some specs https://docs.ai-thinker.com/_media/lora/docs/ra-01sh_specification.pdf ) and not the sx1268.

OK, I described the wrong comparison. I have to set tcxoVoltage to 0 with both mcu's (esp32 and stm32) and this module. With the heltec lora v3, which also should be a ESP32 and a sx1262 the default tcxoVoltage=1.6 works. But OK, maybe other Lora-Hardware!?

The relevant code is very simple. Here the case with try to startReceive():

```
//ESP32
// MOSI pin:   23
// MISO  pin:  19
// SCK pin:  18
#define SX1262_NSS  5
#define SX1262_DIO1 12
#define SX1262_NRST 14
#define SX1262_BUSY 4
#define SX1262_TCXOVOLTAGE 0

void setup() {
  // initialize SX1262 with default settings
  Serial.print(F("[SX1262] Initializing ... "));
  int state = radio.begin(868.0,125.0,7,5,RADIOLIB_SX126X_SYNC_WORD_PRIVATE,10,8,SX1262_TCXOVOLTAGE,false);
  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("success!"));
    u8x8log.print("Init success!\n");
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    u8x8log.print("Init failed!\n");
    while (true);
  }
  radio.setPacketReceivedAction(setFlag);

  // start listening for LoRa packets
  Serial.print(F("[SX1262] Starting to listen ... "));
  state = radio.startReceive();
  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("success!"));
    u8x8log.print("Listening...\n");
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    u8x8log.print("Listen failed!\n");
    while (true);
  }
...
}
```
And that is the output:

```
[SX1262] Initializing ... RLB_DBG: 
RadioLib Info
Version:  "6.6.0.0"
Platform: "ESP32"
Compiled: "Jul 12 2024" "20:41:36"
RLB_SPI: CMDW   80
RLB_SPI: SI             0
RLB_SPI: SO     80      80
RLB_SPI: CMDR   1D      3       20
RLB_SPI: SI                             0       0       0       0       0       0       0       0       0       0       0       0       0       0       0       0       0
RLB_SPI: SO     A2      A2      A2      A2      53      58      31      32      36      31      20      56      32      44      20      32      44      30      32      0
RLB_DBG: Found SX126x: RADIOLIB_SX126X_REG_VERSION_STRING:
RLB_DBG: 0000320 53 58 31 32 36 31 20 56 32 44 20 32 44 30 32 00 | SX1261 V2D 2D02.
RLB_DBG:
RLB_DBG: M      SX126x
RLB_SPI: CMDW   80
RLB_SPI: SI             0
RLB_SPI: SO     FF      FF
RLB_SPI: CMDW   80
RLB_SPI: SI             0
RLB_SPI: SO     A2      A2
RLB_SPI: CMDW   80
RLB_SPI: SI             0
RLB_SPI: SO     A2      A2
RLB_SPI: CMDW   8F
RLB_SPI: SI             0       0
RLB_SPI: SO     A2      A2      A2
RLB_SPI: CMDW   8A
RLB_SPI: SI             1
RLB_SPI: SO     A2      A2
RLB_SPI: CMDW   93
RLB_SPI: SI             20
RLB_SPI: SO     A2      A2
RLB_SPI: CMDW   88
RLB_SPI: SI             3       16      A       0       0       0       0
RLB_SPI: SO     A2      A2      A2      A2      A2      A2      A2      A2
RLB_SPI: CMDW   2
RLB_SPI: SI             43      FF
RLB_SPI: SO     A2      A2      A2
RLB_SPI: CMDW   8
RLB_SPI: SI             0       0       0       0       0       0       0       0
RLB_SPI: SO     A2      A2      A2      A2      A2      A2      A2      A2      A2
RLB_SPI: CMDW   89
RLB_SPI: SI             7F
RLB_SPI: SO     A2      A2
RLB_SPI: CMDR   C0
RLB_SPI: SI             0       0
RLB_SPI: SO     A2      22      22
RLB_SPI: CMDR   11
RLB_SPI: SI             0       0
RLB_SPI: SO     A2      A2      1
RLB_SPI: CMDW   8B
RLB_SPI: SI             9       6       1       0
RLB_SPI: SO     A2      A2      A2      A2      A2
RLB_SPI: CMDR   11
RLB_SPI: SI             0       0
RLB_SPI: SO     A2      A2      1
RLB_SPI: CMDW   D       7       40
RLB_SPI: SI                             14      24
RLB_SPI: SO     A2      A2      A2      A2      A2
RLB_SPI: CMDR   11
RLB_SPI: SI             0       0
RLB_SPI: SO     A2      A2      1
RLB_SPI: CMDR   1D      7       36
RLB_SPI: SI                             0       0
RLB_SPI: SO     A2      A2      A2      A2      D
RLB_SPI: CMDR   C0
RLB_SPI: SI             0       0
RLB_SPI: SO     A2      22      22
RLB_SPI: CMDW   D       7       36
RLB_SPI: SI                             D
RLB_SPI: SO     A2      A2      A2      A2
RLB_SPI: CMDW   8C
RLB_SPI: SI             0       8       0       FF      1       0
RLB_SPI: SO     A2      A2      A2      A2      A2      A2      A2
RLB_SPI: CMDW   96
RLB_SPI: SI             1
RLB_SPI: SO     A2      A2
RLB_SPI: CMDW   D       8       E7
RLB_SPI: SI                             18
RLB_SPI: SO     A2      A2      A2      A2
RLB_SPI: CMDW   9D
RLB_SPI: SI             1
RLB_SPI: SO     A2      A2
RLB_SPI: CMDR   11
RLB_SPI: SI             0       0
RLB_SPI: SO     A2      A2      1
RLB_SPI: CMDR   1D      7       36
RLB_SPI: SI                             0       0
RLB_SPI: SO     A2      A2      A2      A2      D
RLB_SPI: CMDR   C0
RLB_SPI: SI             0       0
RLB_SPI: SO     A2      22      22
RLB_SPI: CMDW   D       7       36
RLB_SPI: SI                             D
RLB_SPI: SO     A2      A2      A2      A2
RLB_SPI: CMDW   8C
RLB_SPI: SI             0       8       0       FF      1       0
RLB_SPI: SO     A2      A2      A2      A2      A2      A2      A2
RLB_SPI: CMDR   11
RLB_SPI: SI             0       0
RLB_SPI: SO     A2      A2      1
RLB_SPI: CMDR   1D      7       36
RLB_SPI: SI                             0       0
RLB_SPI: SO     A2      A2      A2      A2      D
RLB_SPI: CMDR   C0
RLB_SPI: SI             0       0
RLB_SPI: SO     A2      22      22
RLB_SPI: CMDW   D       7       36
RLB_SPI: SI                             D
RLB_SPI: SO     A2      A2      A2      A2
RLB_SPI: CMDW   8C
RLB_SPI: SI             0       8       0       FF      1       0
RLB_SPI: SO     A2      A2      A2      A2      A2      A2      A2
RLB_SPI: CMDR   11
RLB_SPI: SI             0       0
RLB_SPI: SO     A2      A2      1
RLB_SPI: CMDW   8B
RLB_SPI: SI             7       6       1       0
RLB_SPI: SO     A2      A2      A2      A2      A2
RLB_SPI: CMDR   11
RLB_SPI: SI             0       0
RLB_SPI: SO     A2      A2      1
RLB_SPI: CMDW   8B
RLB_SPI: SI             7       4       1       0
RLB_SPI: SO     A2      A2      A2      A2      A2
RLB_SPI: CMDW   98
RLB_SPI: SI             D7      DB
RLB_SPI: SO     A2      A2      A2
RLB_SPI: CMDW   86
RLB_SPI: SI             36      40      0       0
RLB_SPI: SO     A2      A2      A2      A2      A2
RLB_SPI: CMDR   1D      8       D8
RLB_SPI: SI                             0       0
RLB_SPI: SO     A2      A2      A2      A2      C8
RLB_SPI: CMDR   C0
RLB_SPI: SI             0       0
RLB_SPI: SO     A2      22      22
RLB_SPI: CMDW   D       8       D8
RLB_SPI: SI                             DE
RLB_SPI: SO     A2      A2      A2      A2
RLB_SPI: CMDR   1D      8       E7
RLB_SPI: SI                             0       0
RLB_SPI: SO     A2      A2      A2      A2      18
RLB_SPI: CMDR   C0
RLB_SPI: SI             0       0
RLB_SPI: SO     A2      22      22
RLB_SPI: CMDW   95
RLB_SPI: SI             4       7       0       1
RLB_SPI: SO     A2      A2      A2      A2      A2
RLB_SPI: CMDW   8E
RLB_SPI: SI             A       4
RLB_SPI: SO     A2      A2      A2
RLB_SPI: CMDW   D       8       E7
RLB_SPI: SI                             18
RLB_SPI: SO     A2      A2      A2      A2
success!
[SX1262] Starting to listen ... RLB_SPI: CMDW   8
RLB_SPI: SI             2       62      0       2       0       0       0       0
RLB_SPI: SO     FF      FF      FF      FF      FF      FF      FF      FF      FF
failed, code -2
```

And now I tried to startTransmit() and before I didn't check the state in my quick&dirty test. If I check that now, it fails also with error code -1

```
[SX1262] Sending first packet ... RLB_SPI: CMDR 11
RLB_SPI: SI             0       0
RLB_SPI: SO     FF      FF      FF
failed, code -1
```

Do you have an idea how to analyse the problem?

**Reply by jgromes** [reply, marked as answer]
> The ra-01sh seems to be a sx1262

Typo on my part, but it doesn't matter whether this is SX1268 or SX1262, the point about TCXO applies in general to any SX126x.

I see you are using some display, and from the code it looks the failure alway occurs after `u8x8log.print`. Does the display use SPI and is it the same SPI you are using for the LoRa radio?

**Reply by SamLue** [reply]
Sorry!

You were absolutely right. I had a problem with the display. No idea why "state == RADIOLIB_ERR_NONE" then was false, but it was definitly a problem with the display.

Thanks for your help!

**Comment by Nandhini-271** [comment]
I am currently working on multiple two-way communications with the LoRa SX1262 module connect with ESP32, introducing intermittent delays of 2-3 seconds. The first set of two-way communication, involving sending a request and receiving a response, is successfully completed in both transmission and reception modes. However, during the second set of communication, where sensor data is sent and an acknowledgment is expected, the transmitter sends the sensor data but receives its own transmitted data instead of the acknowledgment. Similarly, the receiver, in sensor data receiving mode, receives its own response data.
I have ensured a 2-second delay between transmissions and verified that the receiver is sending the acknowledgment packet. I am unable to identify the source of the issue. Could you please provide a solution? This is an urgent matter.
